### PR TITLE
sonic-moe: v2 ep support

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
## Important: Read before submitting

> **New contributors:** Please open or comment on an issue **before** submitting a
> PR to discuss the change you'd like to make. This helps us align on approach and
> avoids wasted effort on changes we may not be able to merge. Please only create
> a PR once one of the project maintainers agrees on your outlined approach.
>
> PRs of contributors who are not vouched for are automatically closed. Regular
> contributors are added to the vouch list.
>
> The kernels-community repository is for:
>
> 1. Kernels developed by Hugging Face and partners.
> 2. Kernels developed by third parties, that have not been 'kernelized' yet, but
>    are used by Hugging Face projects such as diffusers and transformers.
>
> We cannot accept PRs for the following:
>
> - New kernels that are not in categories (1) and (2).
> - Changes to kernels in category (2), these should be submitted upstream.
>
> For, LLM-generated changes, we prefer that you write your prompt in an issue
> over a PR with LLM-generated changes.

## Related issue

Closes #

## Changes

-

## Testing

-

## Checklist

- [ ] This PR is linked to an issue that was discussed and approved
- [ ] I have tested these changes locally
- [ ] New/changed functionality has test coverage
- LLM disclosure:
  - [ ] I did not use an LLM to create this PR.
  - [ ] I used and LLM for assistance while creating this PR.
  - [ ] This PR was mostly or completely generated by an LLM.

### CODE REVIEW

File: sonic-moe/torch-ext/sonic_moe/functional/reduction_over_k_gather.py [hunk 1]
```diff
@@ -58,6 +58,7 @@ def token_gather_sum_kernel(
     T,
     H: tl.constexpr,
     MAX_K: tl.constexpr,
+    Mtotal: tl.constexpr,  # int32, total rows in x. Doubles as the EP-sentinel marker: M_perm[slot] == Mtotal → skip.
```
Comment (NanoCode012) [inline sonic-moe/torch-ext/sonic_moe/functional/reduction_over_k_gather.py]: Setting it as `tl.constexpr` could cause it to be recompiled unnecessarily? I think it can be an ordinary scalar like T?

File: sonic-moe/torch-ext/sonic_moe/functional/backward.py [hunk 2]
```diff
@@ -208,6 +208,22 @@ def _up_projection_backward_act(
...
+    uninitialized. `s_reverse_scatter_idx` maps every slot either to its own row or to `TK` for a
+    sentinel, so gathering off a buffer with one appended zero row gives each slot its row or a 0: the
+    uninitialized tail is never addressed, and no slot is written twice. Without sentinels it is a
+    plain permutation. Only the general-routing metadata writes the `TK` marker; the bitmatrix path
+    never sees sentinels, so add the marker there too if that changes.
+    """
+    TK = s_reverse_scatter_idx.numel()
+    padded = torch.zeros(TK + 1, dtype=out.dtype, device=out.device)
+    padded[:TK].copy_(ds_scattered)
+    torch.index_select(padded, 0, s_reverse_scatter_idx, out=out.view(-1))
```
Comment (NanoCode012) [inline sonic-moe/torch-ext/sonic_moe/functional/backward.py]: Small nit: Claude recommended this instead to reduce tl launches and reduce allocation. Worth a look?

```diff
  +@triton.jit
  +def _gather_valid_rows_kernel(src_ptr, rev_ptr,
  out_ptr, TK, BLOCK: tl.constexpr):
  +    offs = tl.program_id(0) * BLOCK + tl.arange(0,
  BLOCK)
  +    slot_mask = offs < TK
  +    rev = tl.load(rev_ptr + offs, mask=slot_mask,
  other=TK)
  +    valid = slot_mask & (rev < TK)
  +    val = tl.load(src_ptr + tl.where(valid, rev, 0),
  mask=valid, other=0.0)
  +    tl.store(out_ptr + offs,
  val.to(out_ptr.dtype.element_ty), mask=slot_mask)
  +
  +
   def _gather_valid_rows(ds_scattered,
  s_reverse_scatter_idx, out):
       """..."""
       TK = s_reverse_scatter_idx.numel()
  -    padded = torch.zeros(TK + 1, dtype=out.dtype,
  device=out.device)
  -    padded[:TK].copy_(ds_scattered)
  -    torch.index_select(padded, 0,
  s_reverse_scatter_idx, out=out.view(-1))
  +    BLOCK = 1024
  +    _gather_valid_rows_kernel[(triton.cdiv(TK,
  BLOCK),)](
  +        ds_scattered, s_reverse_scatter_idx, out, TK,
  BLOCK=BLOCK
  +    )
```

File: sonic-moe/torch-ext/sonic_moe/functional/triton_kernels/__init__.py [hunk 3]
```diff
@@ -203,31 +208,40 @@ def _general_metadata_compute_stage2(
...
     token_idx = tl.load(sorted_selected_T_ptr + entry_idx, mask=mask)
 
-    tl.store(s_reverse_scatter_idx_ptr + entry_idx, s_reverse_scatter_val, mask=mask)
+    # Sentinel lanes get `TK` written into s_reverse_scatter_idx as an out-of-range marker, so the
+    # bwd reduction (`token_gather_sum_kernel`) skips them via the `< M` bound check. The output-
+    # indexed tail [sum_valid, TK) of s_scatter_idx / x_gather_idx is left untouched — caller
+    # zero-inits those so downstream reads (topk_scores[s_scatter_idx], reverse scatter) are 0.
+    tl.store(s_reverse_scatter_idx_ptr + entry_idx, tl.where(mask, s_reverse_scatter_val, TK), mask=entry_idx < TK)
     tl.store(s_scatter_idx_ptr + s_reverse_scatter_val, entry_idx, mask=mask)
     tl.store(x_gather_idx_ptr + s_reverse_scatter_val, token_idx, mask=mask)
```
Comment (NanoCode012) [inline sonic-moe/torch-ext/sonic_moe/functional/triton_kernels/__init__.py:246]:
```suggestion
safe_rev = tl.where(mask, s_reverse_scatter_val, 0)
tl.store(s_scatter_idx_ptr + safe_rev, entry_idx, mask=mask)
tl.store(x_gather_idx_ptr + safe_rev, token_idx, mask=mask)
```

Should we add a clamp here? (My suggestion indent format is off here)

**Review by vasqu — DISMISSED** [review]
So we properly mask OOB, let's iterate fast before I'm out 🫡

**Review by NanoCode012 — COMMENTED** [review]
Thanks for the integration update! Found some small points below

File: sonic-moe/torch-ext/sonic_moe/functional/triton_kernels/__init__.py (hunk 3 above)
Comment (IlyasMoutawwakil) [reply sonic-moe/torch-ext/sonic_moe/functional/triton_kernels/__init__.py:246]: why ? I tried to verify this with "compute-sanitizer --tool memcheck" over an EP test with real sentinel routing but it reports "ERROR SUMMARY: 0 errors" without the clamp.

File: sonic-moe/torch-ext/sonic_moe/functional/backward.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply sonic-moe/torch-ext/sonic_moe/functional/backward.py]: yes this has less overhead ! thanks !

File: sonic-moe/torch-ext/sonic_moe/functional/reduction_over_k_gather.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply sonic-moe/torch-ext/sonic_moe/functional/reduction_over_k_gather.py]: yes true ! using a scalar

**Comment by IlyasMoutawwakil** [comment]
Thanks @NanoCode012 🤗

**Review by vasqu — APPROVED** [review]
